docs/docs/aztec/concepts/smart_contracts/functions/function_transforms.md

```diff
@@ -0,0 +1,277 @@
+---
+title: Inner Workings of Functions
+sidebar_position: 3
+tags: [functions]
+---
+
+Below, we go more into depth of what is happening under the hood when you create a function in an Aztec contract. The [next page](./attributes.md) will give you more information about what the attributes are really doing.
+
+
+## Function transformation
+
+When you define a function in an Aztec contract, it undergoes several transformations when it is compiled. These transformations prepare the function for execution. These transformations include:
+
+- [Creating a context for the function](#context-creation)
+- [Handling function inputs](#input-handling)
+- [Processing return values](#return-value-handling)
+- [Computing note hashes and nullifiers](#computing-note-hash-and-nullifier)
+- [Generating function signatures](#function-signature-generation)
+- [Generating contract artifacts](#contract-artifacts)
+
+Let's explore each of these transformations in detail.
+
+## Context creation
+
+Every function in an Aztec contract operates within a specific context which provides some extra information and functionality. This is either a `PrivateContext` or `PublicContext` object, depending on whether it is a private or public function. For private functions, it creates a hash of all input parameters to ensure privacy.
+
+### Private functions
+
+For private functions, the context creation involves hashing all input parameters:
+
+```rust
+let mut args_hasher = ArgsHasher::new();
+// Hash each parameter 
+args_hasher.add(param1);
+args_hasher.add(param2);
+// add all parameters
+
+let mut context = PrivateContext::new(inputs, args_hasher.hash());
+```
+
+This hashing process is important because it is used to verify the function's execution without exposing the input data.
+
+### Public functions
+
+For public functions, context creation is simpler:
+
+```noir
```

Comment:
is noir recognized now?

Reply:
I don't think it is, docusaurus uses prism to fetch languages... Could be worth trying to get Noir into prism tho

Reply:
it aint. was in my own world. changed to rust